### ISSUE TITLE
Improve configuration of multiple PCA963x chips

### ISSUE THREAD

**Issue by Luxovious** [RobTillaart/PCA9634#12]
Hi @RobTillaart,

Currently in your development branch the .config function and its attributes are fully defined in the library. However, defining the mode1_mask /mode2_mask within the library might not be the best solution, as this would lead to the same configuration of all available chips. Could it be usefill to have the input (the modemask) as an argument for the function? The config function could then be constructed similar to the writeN-function. Both mode-registers could be written very efficient with autoincrement with very few commands. If the configuration function is written in a smart way (I don't know how ☹), it could be used by users with structures/arrays. If done right, users could build up an array/struct as a function input for defining and configuring many PCA963x chips at the same time.

Something along the lines:
```cpp
typedef struct
  {		string PCA9634_name;
		int PCA9634_addresses;
		int PCA9634_mode1mask;
		int PCA9634_mode2mask;
  }  PCA9634_define;

PCA9634_define myPCA9634_array [];

myPCA9634_array [20] = (
{myPCA9634_1, 0x40,	PCA9634_MODE1_ALLCALL,				PCA9634_MODE2_TOTEMPOLE | PCA9634_MODE2_STOP | PCA9634_MODE2_INVERT},
{myPCA9634_2, 0x50,	PCA9634_MODE1_ALLCALL | PCA9634_SUBADR(1),	PCA9634_MODE2_TOTEMPOLE | PCA9634_MODE2_STOP | PCA9634_MODE2_INVERT},
{myPCA9634_3, 0x60, ...
);// for 20 units
```
--> now your setup function (modified)
```cpp
PCA9634 myPCA9634_array (string "name"); // create all instances at once
```
and 
```cpp
PCA9634.config (PCA9634_define[address, mode1_mask, mode2_mask]); //configures the chips with efficient I2C bus communication at once.
```
I tried something similar before, however I failed since I'm not good enough of a programmer (yet?).

**Comment by RobTillaart** [comment]
You're trying to make multiple steps at once, that makes such construct often hard to program. 
Also keep in mind that optimizing an initialization - which is typically done only once - is not always efficient. 
It might increase the footprint of the library and not all boards have much memory especially the often used UNO.

That said, I like the idea to add optional parameters to **config()** as a first step.

**Comment by RobTillaart** [comment]
new configure() will look like

```cpp
  void     configure(uint8_t mode1_mask = PCA9634_MODE1_ALLCALL, 
                     uint8_t mode2_mask = PCA9634_MODE2_NONE);
```

implementation
```cpp
void PCA9634::configure(uint8_t mode1_mask, uint8_t mode2_mask)
{
  _data = 0;
  _error = 0;

  setMode1(mode1_mask);
  setMode2(mode2_mask);
}
```

**Comment by Luxovious** [comment]
I adapted the configure() and the begin() functions the way you indicated. Works like a charm:
.h-file
```cpp
bool     begin(uint8_t mode1_mask, uint8_t mode2_mask);
void     configure(uint8_t mode1_mask, uint8_t mode2_mask);
```
.cpp-file
```cpp
bool PCA9634::begin(uint8_t mode1_mask, uint8_t mode2_mask)
{
  _wire->begin();
  if (! isConnected()) return false;
  configure(mode1_mask, mode2_mask);
  return true;
}

void PCA9634::configure(uint8_t mode1_mask, uint8_t mode2_mask)
{
  _data = 0;
  _error = 0;

  setMode1(mode1_mask);
  setMode2(mode2_mask);
}
```

I tested a few different behaviour upon different inputs to begin().
When two uint8_t are used, (plain text or XXX | YYY | ZZZ) it works perfect. When the brackets of begin() are left empty, the compiler complains as two arguments are missing and does not compile. When 0,0 is entered most funcitnos of the chip are turned off. After another configure() or begin() with two arguments the chips works again as expected.

Additionally I recommend a name change in the .h-file:
I found one missleading name and renamed PCA9634_MODE2_STOP to PCA9634_MODE2_ACK, as it is ACK when set to 1 and STOP when 0

**Comment by RobTillaart** [comment]
I will include these into the upcoming 0.2.0 - see https://github.com/RobTillaart/PCA9634/pull/13

**Comment by RobTillaart** [comment]
small diff

```cpp
void     configure(uint8_t mode1_mask, uint8_t mode2_mask);

==>

void     configure(uint8_t mode1_mask = PCA9634_MODE1_ALLCALL, 
                   uint8_t mode2_mask = PCA9634_MODE2_NONE);
```

Added  PCA9634_MODE1_NONE + PCA9634_MODE2_NONE   too (both 0x00)

**Comment by RobTillaart** [comment]
The configure above is in the new 0.2.0 release. (master branch)
Please give it a try.

**Comment by Luxovious** [comment]
Hi @RobTillaart,
I have tested your new release and it does work. In your code you use the begin() function without arguments. The configure() function is called within the begin() function. It is called with the arguments which are placed in the header file. This works perfect when you use a single PCA9634 chip. As soon as you want to use another PCA9634 with different configuration you would have to run the begin()-function as before and an additional configure() function adapting the config of the second PCA9634. Therefore, the configure() function has to be run twice for every PCA9634 that deviates from the configuration in the .h-file.

When using many PCA9634s (10 - 20 will be necessary in my project), this will become very tedious. As I have suggested above, the begin() function could already bear the arguments. It would "forward" those to the configure function. 
I'm very confident that this adapted routine could be used to work with larger configuration arrays and a loop function without the need to double-configure the PCA9634 chips.

I really like your idea to use default values for the configure() function to make sure the code will compile.

Furthermore, I'm happy to write a few words for the readme, regarding the configuration, in case you are interested. Just let me know.

**Comment by RobTillaart** [comment]
Understand what your point is.
Sounds like we have a 0.2.1 release soon.
If it is just the parameters for begin it can be done tonight.

Maybe you could write an example with the subaddressing?

**Comment by Luxovious** [comment]
Sure, I'll do my best.
I'm not sure if I'll be able to finish if tonight. As I'll be on a business trip, I will continue later this week.
Also, I'm going to re-use a lot of code of your previous examples.

**Comment by Luxovious** [comment]
I will create the base of the example today, however I will wait to finalize and test it untill 0.2.1 is released as I will use the "new" begin() function. The current examples should also receive an update. I'm happy to do that as well, afterwards.

**Comment by RobTillaart** [comment]
0.2.1 is in the develop branch - CI build is running

**Comment by RobTillaart** [comment]
@Luxovious 
build succeeded, I will merge into master and will release 0.2.1 after you had a look at it.

**Comment by Luxovious** [comment]
I took a look at it and found one small thing. I mentioned it in the pull request.
I finished the example. It does work. Where do I put that? Another Issue?
I'll be off for today (and most of the week).
Good night.

**Comment by RobTillaart** [comment]
Fixed in version 0.2.1
